app/Connect/Actions/BuildClientPatchDataV2Action.php

```diff
@@ -0,0 +1,359 @@
+     * If the user loads a specialty set hash, always send the specialty game's RP if available.
+     *   Fall back to the core game's RP if there's no specialty game RP.
+     * If the user loads an exclusive set hash, always send the exclusive game's RP if available.
+     *   Fall back to the core game's RP if there's no exclusive game RP.
+     */
+    private function buildRichPresenceData(Game $actualLoadedGame, Game $derivedCoreGame, AchievementSetType $loadedSetType): array
+    {
+        $doesLoadedGameHaveRp =
+            !empty($actualLoadedGame->RichPresencePatch)
+            && !is_null($actualLoadedGame->RichPresencePatch);
+
+        $didUserLoadSpecialtyOrExclusive =
+            $loadedSetType === AchievementSetType::Specialty
+            || $loadedSetType === AchievementSetType::Exclusive;
+
+        if ($doesLoadedGameHaveRp && $didUserLoadSpecialtyOrExclusive) {
+            return [$actualLoadedGame->id, $actualLoadedGame->RichPresencePatch];
+        }
+
+        return [$derivedCoreGame->id, $derivedCoreGame->RichPresencePatch];
+    }
+
+    /**
+     * Calculates the total number of players for the game, which ultimately gets used in
+     * achievement rarity calculations.
+     *
+     * This method adds 1 to the total if the requesting user hasn't played the game yet,
+     * which ensures accurate rarity predictions for when they unlock achievements.
+     *
+     * @param Game $game The game to calculate player count for
+     * @param User|null $user The current user requesting the data
+     *
+     * @return int The total number of players (minimum of 1 to prevent division by zero)
+     */
+    private function calculateGamePlayerCount(Game $game, ?User $user): int
+    {
+        $gamePlayerCount = $game->players_total;
+
+        if ($user) {
+            $hasPlayerGame = PlayerGame::whereUserId($user->id)
+                ->whereGameId($game->id)
+                ->exists();
+
+            if (!$hasPlayerGame) {
+                $gamePlayerCount++;
+            }
+        }
+
+        return max(1, $gamePlayerCount);
+    }
+
+    private function buildIncompatiblePatchData(
+        Game $game,
+        GameHashCompatibility $gameHashCompatibility,
+        ?User $user,
+    ): array {
+        $seeSupportedGameFiles = 'See the Supported Game Files page for this game to find a compatible version.';
+
+        $coreAchievementSet = GameAchievementSet::where('game_id', $game->id)
+            ->core()
+            ->first();
+        $achievementSetId = $coreAchievementSet ? $coreAchievementSet->achievement_set_id : 0;
+
+        return [
+            'Success' => true,
+            'GameId' => VirtualGameIdService::encodeVirtualGameId($game->id, $gameHashCompatibility),
+            'Title' => 'Unsupported Game Version',
+            'ImageIconUrl' => media_asset($game->ImageIcon),
+            'ConsoleId' => $game->ConsoleID,
+            'Sets' => [
```

Comment:
The needs to be another level deeper (there's no array in the response):
```
..."Sets":{"Title":null,...
```

Should be
```
..."Sets":[{"Title":null,...
```

Reply:
Made this change in latest.

Reply:
Not sure I'm following, this is what I see when I make an HTTP request:
![Screenshot 2025-05-17 at 9 28 24 AM](https://github.com/user-attachments/assets/672c4159-833a-4962-abd0-9ba8d74c740e)

Reply:
Moving `leaderboards` fixed this. You can put it back if you want to try to reproduce.

I think that's the last of the obvious things I've encountered. You should be able to write the unit tests now.

Reply:
Ah, that makes a lot of sense. Great, I'll get started on writing tests later this afternoon.